src/Options/Factory.hpp

```diff
@@ -51,16 +51,36 @@ struct print_derived {
   }
 };
 
-template <typename BaseClass>
+template <typename CreatableClasses>
 std::string help_derived() noexcept {
   return "Known Ids:\n" +
-         tmpl::for_each<typename BaseClass::creatable_classes>(
+         tmpl::for_each<CreatableClasses>(
              Factory_detail::print_derived{})
              .value;
 }
 
+// This is for handling legacy code that still uses creatable_classes.
```

Comment:
Is there a benefit to completely disabling `createable_classes`? For things like the boundary corrections it seems much simpler to just list the classes in the base class and have a separate `Factory.hpp` file that does all the includes. I think for things like Events, Triggers, etc. that use the `Registrars` factoring things out is maybe more intuitive, but for the case where all the derived classes are easily listed in the base class it seems like it would actually be less intuitive. Thoughts on this?

Reply:
How about the `Factory.hpp` file both includes everything and defines a typelist with all derived classes? Then you can use that typelist for `call_with_dynamic_type` without going through the metavars.

Reply:
But I still need to list the factory info in the metavars, that's the part I'm trying to avoid because I find the metavars are already a bit difficult to understand

Reply:
Obviously I intended for a full replacement, but I'm not super attached to that.  Just seemed simplest if there was only one way to do it.

In terms of straight lines of code in the executable header, I expect it to be about a wash.  Looking at the StepController commit, excluding the lines added only because this is the first use of the new code, I added an include and a line in the map, and removed a registration line at the bottom.  Things already using a Factory.hpp won't even need the new include.

> How about the `Factory.hpp` file both includes everything and defines a typelist with all derived classes? Then you can use that typelist for `call_with_dynamic_type` without going through the metavars.

I don't see much point in a `Factory.hpp` file if you have to include it in the base class header.  The main point of those is to avoid needing to include all the derived class header files in the base class header to reduce compile times.  Everything can just be put directly into the base class header if this is necessary.

Reply:
> I don't see much point in a Factory.hpp file if you have to include it in the base class header. The main point of those is to avoid needing to include all the derived class header files in the base class header to reduce compile times. Everything can just be put directly into the base class header if this is necessary.

If you have to include the `Factory.hpp` in the base class there's indeed little point to it, though I think @nilsdeppe is referring to the pattern in e.g. [`Evolution/Systems/ScalarWave/BoundaryConditions/BoundaryCondition.hpp`](https://github.com/sxs-collaboration/spectre/blob/develop/src/Evolution/Systems/ScalarWave/BoundaryConditions/BoundaryCondition.hpp) and [`Evolution/Systems/ScalarWave/BoundaryConditions/Factory.hpp`](https://github.com/sxs-collaboration/spectre/blob/develop/src/Evolution/Systems/ScalarWave/BoundaryConditions/Factory.hpp): The base class forward-declares and lists the `creatable_classes`, and the factory file includes the derived headers. There's also [`Evolution/Systems/ScalarWave/BoundaryConditions/RegisterDerivedWithCharm.hpp`](https://github.com/sxs-collaboration/spectre/blob/develop/src/Evolution/Systems/ScalarWave/BoundaryConditions/RegisterDerivedWithCharm.hpp) that bundles the registrations. I think this pattern can be simplified eventually (in particular the charm-registration) but we can keep supporting it until we have a better idea how it plays out.

Reply:
Yea, that's the pattern I was referring to. For the boundary corrections it might be nice to maintain the list of all the available corrections outside of the executable's `Metavariables`. The boundary condition tests need info about what boundary corrections exist because what the boundary condition needs to compute can depend on the boundary correction. This would mean having a typelist like `createable_classes` maintained somewhere in the `BoundaryCorrections` directory would be nice. I'm not sure it really matters whether that be inside the base class, inside the base class header but not in the class, or inside a different header like `Factory.hpp`. Does that make sense?

Reply:
I was thinking something like this.  The list could also be fetched from `Factory.hpp` instead of the metavariables.  I don't think it matters.

<details><summary>7 files changed, 17 insertions(+), 51 deletions(-)</summary>

```patch
diff --git a/src/Evolution/DiscontinuousGalerkin/Actions/BoundaryConditionsImpl.hpp b/src/Evolution/DiscontinuousGalerkin/Actions/BoundaryConditionsImpl.hpp
index b86dc8584..ab2019e62 100644
--- a/src/Evolution/DiscontinuousGalerkin/Actions/BoundaryConditionsImpl.hpp
+++ b/src/Evolution/DiscontinuousGalerkin/Actions/BoundaryConditionsImpl.hpp
@@ -44,6 +44,7 @@
 #include "NumericalAlgorithms/LinearOperators/PartialDerivatives.hpp"
 #include "NumericalAlgorithms/Spectral/Mesh.hpp"
 #include "NumericalAlgorithms/Spectral/Spectral.hpp"
+#include "Parallel/MetavariablesTag.hpp"
 #include "Time/Tags.hpp"
 #include "Utilities/ErrorHandling/Assert.hpp"
 #include "Utilities/ErrorHandling/Error.hpp"
@@ -667,7 +668,10 @@ void apply_boundary_conditions_on_all_external_faces(
     const Variables<detail::get_primitive_vars_tags_from_system<System>>* const
         primitive_vars) noexcept {
   using derived_boundary_conditions = tmpl::remove_if<
-      typename System::boundary_conditions_base::creatable_classes,
+      tmpl::at<
+          typename std::decay_t<decltype(
+              db::get<Parallel::Tags::Metavariables>(*box))>::factory_classes,
+          typename System::boundary_conditions_base>,
       tmpl::or_<
           std::is_base_of<domain::BoundaryConditions::MarkAsPeriodic, tmpl::_1>,
           std::is_base_of<domain::BoundaryConditions::detail::MarkAsNone,
diff --git a/src/Evolution/Executables/ScalarWave/EvolveScalarWave.hpp b/src/Evolution/Executables/ScalarWave/EvolveScalarWave.hpp
index bb073ed8e..538c2fc24 100644
--- a/src/Evolution/Executables/ScalarWave/EvolveScalarWave.hpp
+++ b/src/Evolution/Executables/ScalarWave/EvolveScalarWave.hpp
@@ -21,8 +21,8 @@
 #include "Evolution/Initialization/Evolution.hpp"
 #include "Evolution/Initialization/NonconservativeSystem.hpp"
 #include "Evolution/Initialization/SetVariables.hpp"
+#include "Evolution/Systems/ScalarWave/BoundaryConditions/BoundaryCondition.hpp"
 #include "Evolution/Systems/ScalarWave/BoundaryConditions/Factory.hpp"
-#include "Evolution/Systems/ScalarWave/BoundaryConditions/RegisterDerivedWithCharm.hpp"
 #include "Evolution/Systems/ScalarWave/BoundaryCorrections/Factory.hpp"
 #include "Evolution/Systems/ScalarWave/BoundaryCorrections/RegisterDerived.hpp"
 #include "Evolution/Systems/ScalarWave/Equations.hpp"
@@ -152,6 +152,8 @@ struct EvolutionMetavars {
       typename Event<events>::creatable_classes>;
 
   using factory_classes = tmpl::map<
+      tmpl::pair<ScalarWave::BoundaryConditions::BoundaryCondition<Dim>,
+                 ScalarWave::BoundaryConditions::creatable_classes<Dim>>,
       tmpl::pair<StepController, tmpl::list<StepControllers::BinaryFraction>>>;
 
   // The scalar wave system generally does not require filtering, except
@@ -306,7 +308,6 @@ static const std::vector<void (*)()> charm_init_node_funcs{
     &domain::creators::register_derived_with_charm,
     &domain::creators::time_dependence::register_derived_with_charm,
     &domain::FunctionsOfTime::register_derived_with_charm,
-    &ScalarWave::BoundaryConditions::register_derived_with_charm,
     &ScalarWave::BoundaryCorrections::register_derived_with_charm,
     &Parallel::register_derived_classes_with_charm<
         Event<metavariables::events>>,
diff --git a/src/Evolution/Systems/ScalarWave/BoundaryConditions/BoundaryCondition.hpp b/src/Evolution/Systems/ScalarWave/BoundaryConditions/BoundaryCondition.hpp
index 82c8131b9..e357061cd 100644
--- a/src/Evolution/Systems/ScalarWave/BoundaryConditions/BoundaryCondition.hpp
+++ b/src/Evolution/Systems/ScalarWave/BoundaryConditions/BoundaryCondition.hpp
@@ -9,29 +9,12 @@
 #include "Domain/BoundaryConditions/Periodic.hpp"
 #include "Utilities/TMPL.hpp"
 
-/// \cond
-namespace ScalarWave::BoundaryConditions {
-template <size_t Dim>
-class ConstraintPreservingSphericalRadiation;
-template <size_t Dim>
-class DirichletAnalytic;
-template <size_t Dim>
-class SphericalRadiation;
-}  // namespace ScalarWave::BoundaryConditions
-/// \endcond
-
 /// \brief Boundary conditions for the scalar wave system
 namespace ScalarWave::BoundaryConditions {
 /// \brief The base class off of which all boundary conditions must inherit
 template <size_t Dim>
 class BoundaryCondition : public domain::BoundaryConditions::BoundaryCondition {
  public:
-  using creatable_classes =
-      tmpl::list<ConstraintPreservingSphericalRadiation<Dim>,
-                 DirichletAnalytic<Dim>,
-                 domain::BoundaryConditions::Periodic<BoundaryCondition<Dim>>,
-                 SphericalRadiation<Dim>>;
-
   BoundaryCondition() = default;
   BoundaryCondition(BoundaryCondition&&) noexcept = default;
   BoundaryCondition& operator=(BoundaryCondition&&) noexcept = default;
diff --git a/src/Evolution/Systems/ScalarWave/BoundaryConditions/CMakeLists.txt b/src/Evolution/Systems/ScalarWave/BoundaryConditions/CMakeLists.txt
index 7a61fa0fe..f213298b9 100644
--- a/src/Evolution/Systems/ScalarWave/BoundaryConditions/CMakeLists.txt
+++ b/src/Evolution/Systems/ScalarWave/BoundaryConditions/CMakeLists.txt
@@ -7,7 +7,6 @@ spectre_target_sources(
   BoundaryCondition.cpp
   ConstraintPreservingSphericalRadiation.cpp
   DirichletAnalytic.cpp
-  RegisterDerivedWithCharm.cpp
   SphericalRadiation.cpp
   )
 
@@ -19,6 +18,5 @@ spectre_target_headers(
   ConstraintPreservingSphericalRadiation.hpp
   DirichletAnalytic.hpp
   Factory.hpp
-  RegisterDerivedWithCharm.hpp
   SphericalRadiation.hpp
   )
diff --git a/src/Evolution/Systems/ScalarWave/BoundaryConditions/Factory.hpp b/src/Evolution/Systems/ScalarWave/BoundaryConditions/Factory.hpp
index 54538febd..e00e8bc9c 100644
--- a/src/Evolution/Systems/ScalarWave/BoundaryConditions/Factory.hpp
+++ b/src/Evolution/Systems/ScalarWave/BoundaryConditions/Factory.hpp
@@ -7,3 +7,12 @@
 #include "Evolution/Systems/ScalarWave/BoundaryConditions/ConstraintPreservingSphericalRadiation.hpp"
 #include "Evolution/Systems/ScalarWave/BoundaryConditions/DirichletAnalytic.hpp"
 #include "Evolution/Systems/ScalarWave/BoundaryConditions/SphericalRadiation.hpp"
+
+namespace ScalarWave::BoundaryConditions {
+template <size_t Dim>
+using creatable_classes =
+    tmpl::list<ConstraintPreservingSphericalRadiation<Dim>,
+               DirichletAnalytic<Dim>,
+               domain::BoundaryConditions::Periodic<BoundaryCondition<Dim>>,
+               SphericalRadiation<Dim>>;
+}  // namespace ScalarWave::BoundaryConditions
diff --git a/src/Evolution/Systems/ScalarWave/BoundaryConditions/RegisterDerivedWithCharm.cpp b/src/Evolution/Systems/ScalarWave/BoundaryConditions/RegisterDerivedWithCharm.cpp
deleted file mode 100644
index 28e26fff2..000000000
--- a/src/Evolution/Systems/ScalarWave/BoundaryConditions/RegisterDerivedWithCharm.cpp
+++ /dev/null
@@ -1,21 +0,0 @@
-// Distributed under the MIT License.
-// See LICENSE.txt for details.
-
-#include "Evolution/Systems/ScalarWave/BoundaryConditions/RegisterDerivedWithCharm.hpp"
-
-#include "Evolution/Systems/ScalarWave/BoundaryConditions/BoundaryCondition.hpp"
-#include "Evolution/Systems/ScalarWave/BoundaryConditions/ConstraintPreservingSphericalRadiation.hpp"
-#include "Evolution/Systems/ScalarWave/BoundaryConditions/DirichletAnalytic.hpp"
-#include "Evolution/Systems/ScalarWave/BoundaryConditions/SphericalRadiation.hpp"
-#include "Parallel/RegisterDerivedClassesWithCharm.hpp"
-
-namespace ScalarWave::BoundaryConditions {
-void register_derived_with_charm() noexcept {
-  Parallel::register_classes_in_list<
-      typename BoundaryCondition<1>::creatable_classes>();
-  Parallel::register_classes_in_list<
-      typename BoundaryCondition<2>::creatable_classes>();
-  Parallel::register_classes_in_list<
-      typename BoundaryCondition<3>::creatable_classes>();
-}
-}  // namespace ScalarWave::BoundaryConditions
diff --git a/src/Evolution/Systems/ScalarWave/BoundaryConditions/RegisterDerivedWithCharm.hpp b/src/Evolution/Systems/ScalarWave/BoundaryConditions/RegisterDerivedWithCharm.hpp
deleted file mode 100644
index 47c4ae799..000000000
--- a/src/Evolution/Systems/ScalarWave/BoundaryConditions/RegisterDerivedWithCharm.hpp
+++ /dev/null
@@ -1,8 +0,0 @@
-// Distributed under the MIT License.
-// See LICENSE.txt for details.
-
-#pragma once
-
-namespace ScalarWave::BoundaryConditions {
-void register_derived_with_charm() noexcept;
-}  // namespace ScalarWave::BoundaryConditions
```
</details>

Reply:
Okay, yes that looks good to me. Thanks for sharing that and making the change in this PR :)